In [86]:
# importando as bibliotecas terceiros
import pandas as pd
import numpy as np
import seaborn as sns

In [87]:
tabela_test = pd.read_csv("C:\\Analise_de_DADOS\\Projeto_salles_loja\\test.csv")
tabela_train = pd.read_csv("C:\\Analise_de_DADOS\\Projeto_salles_loja\\train.csv")
store_sales =  pd.read_csv("C:\\Analise_de_DADOS\\Projeto_salles_loja\\stores.csv")
transaction =  pd.read_csv("C:\\Analise_de_DADOS\\Projeto_salles_loja\\transactions.csv")
holidays_events = pd.read_csv("C:\\Analise_de_DADOS\\Projeto_salles_loja\\holidays_events.csv")
oil = pd.read_csv("C:\\Analise_de_DADOS\\Projeto_salles_loja\\oil.csv")
submission =pd.read_csv("C:\\Analise_de_DADOS\\Projeto_salles_loja\\sample_submission.csv")


In [88]:
tabela_test

id        date  store_nbr                      family  onpromotion
0      3000888  2017-08-16          1                  AUTOMOTIVE            0
1      3000889  2017-08-16          1                   BABY CARE            0
2      3000890  2017-08-16          1                      BEAUTY            2
3      3000891  2017-08-16          1                   BEVERAGES           20
4      3000892  2017-08-16          1                       BOOKS            0
...        ...         ...        ...                         ...          ...
28507  3029395  2017-08-31          9                     POULTRY            1
28508  3029396  2017-08-31          9              PREPARED FOODS            0
28509  3029397  2017-08-31          9                     PRODUCE            1
28510  3029398  2017-08-31          9  SCHOOL AND OFFICE SUPPLIES            9
28511  3029399  2017-08-31          9                     SEAFOOD            0

[28512 rows x 5 columns]

In [89]:
loja_salles = pd.concat([tabela_train,tabela_test],axis=0).reset_index(drop=True)

In [90]:
loja_salles

id        date  store_nbr                      family  sales   
0              0  2013-01-01          1                  AUTOMOTIVE    0.0  \
1              1  2013-01-01          1                   BABY CARE    0.0   
2              2  2013-01-01          1                      BEAUTY    0.0   
3              3  2013-01-01          1                   BEVERAGES    0.0   
4              4  2013-01-01          1                       BOOKS    0.0   
...          ...         ...        ...                         ...    ...   
3029395  3029395  2017-08-31          9                     POULTRY    NaN   
3029396  3029396  2017-08-31          9              PREPARED FOODS    NaN   
3029397  3029397  2017-08-31          9                     PRODUCE    NaN   
3029398  3029398  2017-08-31          9  SCHOOL AND OFFICE SUPPLIES    NaN   
3029399  3029399  2017-08-31          9                     SEAFOOD    NaN   

         onpromotion  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0  
...              ...  
3029395            1  
3029396            0  
3029397            1  
3029398            9  
3029399            0  

[3029400 rows x 6 columns]

In [91]:
loja_salles.drop('id',axis=1,inplace=True)

In [92]:
loja_salles

date  store_nbr                      family  sales  onpromotion
0        2013-01-01          1                  AUTOMOTIVE    0.0            0
1        2013-01-01          1                   BABY CARE    0.0            0
2        2013-01-01          1                      BEAUTY    0.0            0
3        2013-01-01          1                   BEVERAGES    0.0            0
4        2013-01-01          1                       BOOKS    0.0            0
...             ...        ...                         ...    ...          ...
3029395  2017-08-31          9                     POULTRY    NaN            1
3029396  2017-08-31          9              PREPARED FOODS    NaN            0
3029397  2017-08-31          9                     PRODUCE    NaN            1
3029398  2017-08-31          9  SCHOOL AND OFFICE SUPPLIES    NaN            9
3029399  2017-08-31          9                     SEAFOOD    NaN            0

[3029400 rows x 5 columns]

In [93]:
loja_salles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3029400 entries, 0 to 3029399
Data columns (total 5 columns):
 #   Column       Dtype  
---  ------       -----  
 0   date         object 
 1   store_nbr    int64  
 2   family       object 
 3   sales        float64
 4   onpromotion  int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 115.6+ MB


In [94]:
# Mudar coluna Date para tipo Date
# separ datas, dia, mês, ano 
# store_nbr para object(string) --- depois de juntar com a tabela store_sales
#



In [95]:
# Juntando tabelas atraves do metodo merge
loja_salles2 = pd.merge(loja_salles,store_sales,on='store_nbr', how='inner')

In [96]:
loja_salles2

date  store_nbr                      family  sales   
0        2013-01-01          1                  AUTOMOTIVE    0.0  \
1        2013-01-01          1                   BABY CARE    0.0   
2        2013-01-01          1                      BEAUTY    0.0   
3        2013-01-01          1                   BEVERAGES    0.0   
4        2013-01-01          1                       BOOKS    0.0   
...             ...        ...                         ...    ...   
3029395  2017-08-31          9                     POULTRY    NaN   
3029396  2017-08-31          9              PREPARED FOODS    NaN   
3029397  2017-08-31          9                     PRODUCE    NaN   
3029398  2017-08-31          9  SCHOOL AND OFFICE SUPPLIES    NaN   
3029399  2017-08-31          9                     SEAFOOD    NaN   

         onpromotion   city      state type  cluster  
0                  0  Quito  Pichincha    D       13  
1                  0  Quito  Pichincha    D       13  
2                  0  Quito  Pichincha    D       13  
3                  0  Quito  Pichincha    D       13  
4                  0  Quito  Pichincha    D       13  
...              ...    ...        ...  ...      ...  
3029395            1  Quito  Pichincha    B        6  
3029396            0  Quito  Pichincha    B        6  
3029397            1  Quito  Pichincha    B        6  
3029398            9  Quito  Pichincha    B        6  
3029399            0  Quito  Pichincha    B        6  

[3029400 rows x 9 columns]

In [97]:
# Junção de mais uma tabela, de acordo com suas colunas incomuns(Date e Store_nbr)
loja_salles3 = pd.merge(loja_salles2,transaction,on=['date','store_nbr'],how='outer')

In [98]:
# existem 274.296 linhas nulas na coluna transactions
loja_salles3[loja_salles3['transactions'].isnull()].count()

date            274296
store_nbr       274296
family          274296
sales           245784
onpromotion     274296
city            274296
state           274296
type            274296
cluster         274296
transactions         0
dtype: int64

In [99]:
loja_salles3[loja_salles3['date'] == '2012-03-02']

Empty DataFrame
Columns: [date, store_nbr, family, sales, onpromotion, city, state, type, cluster, transactions]
Index: []

In [100]:
loja_salles3

date  store_nbr                      family  sales   
0        2013-01-01          1                  AUTOMOTIVE    0.0  \
1        2013-01-01          1                   BABY CARE    0.0   
2        2013-01-01          1                      BEAUTY    0.0   
3        2013-01-01          1                   BEVERAGES    0.0   
4        2013-01-01          1                       BOOKS    0.0   
...             ...        ...                         ...    ...   
3029395  2017-08-31          9                     POULTRY    NaN   
3029396  2017-08-31          9              PREPARED FOODS    NaN   
3029397  2017-08-31          9                     PRODUCE    NaN   
3029398  2017-08-31          9  SCHOOL AND OFFICE SUPPLIES    NaN   
3029399  2017-08-31          9                     SEAFOOD    NaN   

         onpromotion   city      state type  cluster  transactions  
0                  0  Quito  Pichincha    D       13           NaN  
1                  0  Quito  Pichincha    D       13           NaN  
2                  0  Quito  Pichincha    D       13           NaN  
3                  0  Quito  Pichincha    D       13           NaN  
4                  0  Quito  Pichincha    D       13           NaN  
...              ...    ...        ...  ...      ...           ...  
3029395            1  Quito  Pichincha    B        6           NaN  
3029396            0  Quito  Pichincha    B        6           NaN  
3029397            1  Quito  Pichincha    B        6           NaN  
3029398            9  Quito  Pichincha    B        6           NaN  
3029399            0  Quito  Pichincha    B        6           NaN  

[3029400 rows x 10 columns]

# Entendendo a Tabela Holidays

In [101]:
holidays_events['type'].unique()

array(['Holiday', 'Transfer', 'Additional', 'Bridge', 'Work Day', 'Event'],
      dtype=object)

In [102]:
holidays_events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   date         350 non-null    object
 1   type         350 non-null    object
 2   locale       350 non-null    object
 3   locale_name  350 non-null    object
 4   description  350 non-null    object
 5   transferred  350 non-null    bool  
dtypes: bool(1), object(5)
memory usage: 14.1+ KB


Logo abaixo é verificado que na coluna transferred , existe duas opções booleans, logo se a opção é verdadeira, siginifica que ali era um feriado, mas foi transferido para outro dia, logo não é mais feriado mas um dia comum. e se estiver false, significa que aquele se tornou o novo feriado, logo, aquele não é um dia comum.

Resumindo, precisamos olhar para duas colunas, a coluna transferred e a coluna type. se na coluna type estiver escrito holiday, aquele é o feriado, mas se na coluna transferred dessa mesma linha tiver escrito True, significa que aquele dia não é mais feriado , pois foi transferido. 

Quando encontramos a coluna type escrito Transfer, significa que aquele dia recebeu o feriado de outro dia, logo se tornou feriado. 

|TYPE|         |TRANSFERRED|      |RESULTADO|


|holiday|  +   |True|          =  |NÃO É FERIADO| 


|holiday|  +   |False|         =  |É FERIADO|


|Transfer| +   |False|         =  |É FERIADO| 



Verifiquei tambem que, a coluna Locale, indica se o feriado foi em escala nacional ou local... isto é, se teve efeito em todos ou somente em algumas cidades. logo os NATIONAL, são colocados em todas as lojas. 


Regional é para as regiões, norte, sul...

locale é somente para aquele cidade.

In [103]:
holidays_events[((holidays_events['description'].str.contains('Guayaquil')) & (holidays_events['date'].isin(['2012-10-12','2012-10-09'])))]

date      type    locale locale_name   
19  2012-10-09   Holiday  National     Ecuador  \
20  2012-10-12  Transfer  National     Ecuador   

                            description  transferred  
19           Independencia de Guayaquil         True  
20  Traslado Independencia de Guayaquil        False

In [104]:
holidays_events[holidays_events['date'].duplicated()]

date        type    locale    locale_name   
8    2012-06-25     Holiday     Local      Latacunga  \
9    2012-06-25     Holiday     Local        Machala   
11   2012-07-03     Holiday     Local      El Carmen   
33   2012-12-22  Additional  National        Ecuador   
36   2012-12-24  Additional  National        Ecuador   
40   2012-12-31  Additional  National        Ecuador   
55   2013-05-12       Event  National        Ecuador   
59   2013-06-25     Holiday     Local        Machala   
60   2013-06-25     Holiday     Local      Latacunga   
62   2013-07-03     Holiday     Local  Santo Domingo   
86   2013-12-22     Holiday     Local        Salinas   
111  2014-06-25     Holiday     Local        Machala   
112  2014-06-25     Holiday  Regional       Imbabura   
113  2014-06-25       Event  National        Ecuador   
119  2014-07-03     Holiday     Local  Santo Domingo   
152  2014-12-22  Additional  National        Ecuador   
157  2014-12-26  Additional  National        Ecuador   
177  2015-06-25     Holiday  Regional       Imbabura   
178  2015-06-25     Holiday     Local      Latacunga   
180  2015-07-03     Holiday     Local  Santo Domingo   
205  2015-12-22     Holiday     Local        Salinas   
225  2016-04-21       Event  National        Ecuador   
236  2016-05-01       Event  National        Ecuador   
243  2016-05-07       Event  National        Ecuador   
245  2016-05-08       Event  National        Ecuador   
250  2016-05-12       Event  National        Ecuador   
259  2016-06-25     Holiday  Regional       Imbabura   
260  2016-06-25     Holiday     Local      Latacunga   
262  2016-07-03     Holiday     Local  Santo Domingo   
265  2016-07-24    Transfer     Local      Guayaquil   
283  2016-11-12    Work Day  National        Ecuador   
291  2016-12-22     Holiday     Local        Salinas   
306  2017-04-14     Holiday  National        Ecuador   
316  2017-06-25     Holiday     Local      Latacunga   
317  2017-06-25     Holiday     Local        Machala   
319  2017-07-03     Holiday     Local  Santo Domingo   
342  2017-12-08    Transfer     Local          Quito   
345  2017-12-22  Additional  National        Ecuador   

                                   description  transferred  
8                   Cantonizacion de Latacunga        False  
9                         Fundacion de Machala        False  
11                  Cantonizacion de El Carmen        False  
33                                   Navidad-3        False  
36                                   Navidad-1        False  
40                        Primer dia del ano-1        False  
55                             Dia de la Madre        False  
59                        Fundacion de Machala        False  
60                  Cantonizacion de Latacunga        False  
62                  Fundacion de Santo Domingo        False  
86                    Cantonizacion de Salinas        False  
111                       Fundacion de Machala        False  
112              Provincializacion de Imbabura        False  
113  Mundial de futbol Brasil: Ecuador-Francia        False  
119                 Fundacion de Santo Domingo        False  
152                                  Navidad-3        False  
157                                  Navidad+1        False  
177              Provincializacion de Imbabura        False  
178                 Cantonizacion de Latacunga        False  
180                 Fundacion de Santo Domingo        False  
205                   Cantonizacion de Salinas        False  
225                         Terremoto Manabi+5        False  
236                        Terremoto Manabi+15        False  
243                        Terremoto Manabi+21        False  
245                            Dia de la Madre        False  
250                        Terremoto Manabi+26        False  
259              Provincializacion de Imbabura        False  
260                 Cantonizacion de Latacunga        False  
262                 Fundacio

In [105]:
loja_salles3[loja_salles3['city'] == 'Machala']

date  store_nbr                      family  sales   
1907400  2013-01-01         40                  AUTOMOTIVE    0.0  \
1907401  2013-01-01         40                   BABY CARE    0.0   
1907402  2013-01-01         40                      BEAUTY    0.0   
1907403  2013-01-01         40                   BEVERAGES    0.0   
1907404  2013-01-01         40                       BOOKS    0.0   
...             ...        ...                         ...    ...   
2019595  2017-08-31         41                     POULTRY    NaN   
2019596  2017-08-31         41              PREPARED FOODS    NaN   
2019597  2017-08-31         41                     PRODUCE    NaN   
2019598  2017-08-31         41  SCHOOL AND OFFICE SUPPLIES    NaN   
2019599  2017-08-31         41                     SEAFOOD    NaN   

         onpromotion     city   state type  cluster  transactions  
1907400            0  Machala  El Oro    C        3           NaN  
1907401            0  Machala  El Oro    C        3           NaN  
1907402            0  Machala  El Oro    C        3           NaN  
1907403            0  Machala  El Oro    C        3           NaN  
1907404            0  Machala  El Oro    C        3           NaN  
...              ...      ...     ...  ...      ...           ...  
2019595            0  Machala  El Oro    D        4           NaN  
2019596            0  Machala  El Oro    D        4           NaN  
2019597            0  Machala  El Oro    D        4           NaN  
2019598            0  Machala  El Oro    D        4           NaN  
2019599            0  Machala  El Oro    D        4           NaN  

[112200 rows x 10 columns]

In [106]:
holidays_events[holidays_events['date'] == '2012-06-25']

date     type    locale locale_name                    description   
7  2012-06-25  Holiday  Regional    Imbabura  Provincializacion de Imbabura  \
8  2012-06-25  Holiday     Local   Latacunga     Cantonizacion de Latacunga   
9  2012-06-25  Holiday     Local     Machala           Fundacion de Machala   

   transferred  
7        False  
8        False  
9        False

Transformando coluna de date no formato de dataTime, com o intuito de mesclar a coluna holidays atráves da data( já que é a unica coisa em comum neles alem da localização)

In [107]:
loja_salles3['date']= pd.to_datetime(loja_salles3['date'])

In [108]:
loja_salles3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3029400 entries, 0 to 3029399
Data columns (total 10 columns):
 #   Column        Dtype         
---  ------        -----         
 0   date          datetime64[ns]
 1   store_nbr     int64         
 2   family        object        
 3   sales         float64       
 4   onpromotion   int64         
 5   city          object        
 6   state         object        
 7   type          object        
 8   cluster       int64         
 9   transactions  float64       
dtypes: datetime64[ns](1), float64(2), int64(3), object(4)
memory usage: 231.1+ MB


Tranfomando coluna de datas do holidays também

In [109]:
holidays_events['date'] = pd.to_datetime(holidays_events['date'])

In [110]:
holidays_events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   date         350 non-null    datetime64[ns]
 1   type         350 non-null    object        
 2   locale       350 non-null    object        
 3   locale_name  350 non-null    object        
 4   description  350 non-null    object        
 5   transferred  350 non-null    bool          
dtypes: bool(1), datetime64[ns](1), object(4)
memory usage: 14.1+ KB


In [111]:
loja_salles3

date  store_nbr                      family  sales  onpromotion   
0       2013-01-01          1                  AUTOMOTIVE    0.0            0  \
1       2013-01-01          1                   BABY CARE    0.0            0   
2       2013-01-01          1                      BEAUTY    0.0            0   
3       2013-01-01          1                   BEVERAGES    0.0            0   
4       2013-01-01          1                       BOOKS    0.0            0   
...            ...        ...                         ...    ...          ...   
3029395 2017-08-31          9                     POULTRY    NaN            1   
3029396 2017-08-31          9              PREPARED FOODS    NaN            0   
3029397 2017-08-31          9                     PRODUCE    NaN            1   
3029398 2017-08-31          9  SCHOOL AND OFFICE SUPPLIES    NaN            9   
3029399 2017-08-31          9                     SEAFOOD    NaN            0   

          city      state type  cluster  transactions  
0        Quito  Pichincha    D       13           NaN  
1        Quito  Pichincha    D       13           NaN  
2        Quito  Pichincha    D       13           NaN  
3        Quito  Pichincha    D       13           NaN  
4        Quito  Pichincha    D       13           NaN  
...        ...        ...  ...      ...           ...  
3029395  Quito  Pichincha    B        6           NaN  
3029396  Quito  Pichincha    B        6           NaN  
3029397  Quito  Pichincha    B        6           NaN  
3029398  Quito  Pichincha    B        6           NaN  
3029399  Quito  Pichincha    B        6           NaN  

[3029400 rows x 10 columns]

In [112]:
holidays_events['locale'].unique()

array(['Local', 'Regional', 'National'], dtype=object)

In [113]:
# Renomeando para não dar problemas com o type do dataframe de vendas, já que são dados diferentes
holidays_events.rename(columns={'type':'Type_holidays'},inplace=True)

In [114]:
dic_holidays = {'Holiday':'Holidays','Transfer':'Holidays', 'Additional':'Holidays', 'Bridge':'Holidays','Event': 'Holidays',
                'Work Day':'Work Day'}


In [115]:
# Todos os type_holidays que possuem 0 são dias que estou considerando como feriado. 

feriado= holidays_events[['date','locale_name','locale','Type_holidays','transferred']]
feriado['Type_holidays'] = feriado['Type_holidays'].map(dic_holidays).fillna(feriado['Type_holidays'])
feriado['Type_holidays'] = feriado['Type_holidays'].astype('category').cat.codes
feriado['transferred'] = feriado['transferred'].astype('category').cat.codes
feriado['locale'] = feriado['locale'].astype('category').cat.codes


C:\Users\Thiago\AppData\Local\Temp\ipykernel_1480\462106620.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  feriado['Type_holidays'] = feriado['Type_holidays'].map(dic_holidays).fillna(feriado['Type_holidays'])
C:\Users\Thiago\AppData\Local\Temp\ipykernel_1480\462106620.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  feriado['Type_holidays'] = feriado['Type_holidays'].astype('category').cat.codes
C:\Users\Thiago\AppData\Local\Temp\ipykernel_1480\462106620.py:6: SettingWithCopyWarning: 
A value is tr

In [116]:
holidays_events[holidays_events['Type_holidays'] == 'Work Day']

date Type_holidays    locale locale_name   
42  2013-01-05      Work Day  National     Ecuador  \
43  2013-01-12      Work Day  National     Ecuador   
149 2014-12-20      Work Day  National     Ecuador   
161 2015-01-10      Work Day  National     Ecuador   
283 2016-11-12      Work Day  National     Ecuador   

                            description  transferred  
42              Recupero puente Navidad        False  
43   Recupero puente primer dia del ano        False  
149             Recupero Puente Navidad        False  
161  Recupero Puente Primer dia del ano        False  
283     Recupero Puente Dia de Difuntos        False

In [117]:
feriado[feriado['Type_holidays'] == 1]

date locale_name  locale  Type_holidays  transferred
42  2013-01-05     Ecuador       1              1            0
43  2013-01-12     Ecuador       1              1            0
149 2014-12-20     Ecuador       1              1            0
161 2015-01-10     Ecuador       1              1            0
283 2016-11-12     Ecuador       1              1            0

In [118]:
feriado['Type_holidays'].unique()

array([0, 1], dtype=int8)

In [119]:
feriado[feriado['date'] == '2013-05-12']

date locale_name  locale  Type_holidays  transferred
54 2013-05-12        Puyo       0              0            0
55 2013-05-12     Ecuador       1              0            0

In [120]:
# Separando por localidade  
Regional = feriado[feriado['locale'] == 2]
Local = feriado[feriado['locale'] == 0]
National = feriado[feriado['locale'] == 1 ]
National

date locale_name  locale  Type_holidays  transferred
14  2012-08-10     Ecuador       1              0            0
19  2012-10-09     Ecuador       1              0            1
20  2012-10-12     Ecuador       1              0            0
21  2012-11-02     Ecuador       1              0            0
22  2012-11-03     Ecuador       1              0            0
..         ...         ...     ...            ...          ...
345 2017-12-22     Ecuador       1              0            0
346 2017-12-23     Ecuador       1              0            0
347 2017-12-24     Ecuador       1              0            0
348 2017-12-25     Ecuador       1              0            0
349 2017-12-26     Ecuador       1              0            0

[174 rows x 5 columns]

In [121]:
feriado_local = Local.rename(columns={'transferred':'transferred_local','Type_holidays':'type_local','locale':'locale_local'})
feriado_Regional = Regional.rename(columns={'transferred':'transferred_regional','Type_holidays':'type_regional','locale':'locale_regional'})
feriado_National = National.rename(columns={'transferred':'transferred_national','Type_holidays':'type_national','locale':'locale_national'})

In [122]:
loja_salles4 = loja_salles3.merge(feriado_local, left_on=['date', 'city'], right_on=['date', 'locale_name'], how='left')

In [123]:
loja_salles5 = loja_salles4.merge(feriado_Regional, left_on=['date', 'state'], right_on=['date', 'locale_name'], how='left')


In [124]:
loja_salles5

date  store_nbr                      family  sales  onpromotion   
0       2013-01-01          1                  AUTOMOTIVE    0.0            0  \
1       2013-01-01          1                   BABY CARE    0.0            0   
2       2013-01-01          1                      BEAUTY    0.0            0   
3       2013-01-01          1                   BEVERAGES    0.0            0   
4       2013-01-01          1                       BOOKS    0.0            0   
...            ...        ...                         ...    ...          ...   
3029659 2017-08-31          9                     POULTRY    NaN            1   
3029660 2017-08-31          9              PREPARED FOODS    NaN            0   
3029661 2017-08-31          9                     PRODUCE    NaN            1   
3029662 2017-08-31          9  SCHOOL AND OFFICE SUPPLIES    NaN            9   
3029663 2017-08-31          9                     SEAFOOD    NaN            0   

          city      state type  cluster  transactions locale_name_x   
0        Quito  Pichincha    D       13           NaN           NaN  \
1        Quito  Pichincha    D       13           NaN           NaN   
2        Quito  Pichincha    D       13           NaN           NaN   
3        Quito  Pichincha    D       13           NaN           NaN   
4        Quito  Pichincha    D       13           NaN           NaN   
...        ...        ...  ...      ...           ...           ...   
3029659  Quito  Pichincha    B        6           NaN           NaN   
3029660  Quito  Pichincha    B        6           NaN           NaN   
3029661  Quito  Pichincha    B        6           NaN           NaN   
3029662  Quito  Pichincha    B        6           NaN           NaN   
3029663  Quito  Pichincha    B        6           NaN           NaN   

         locale_local  type_local  transferred_local locale_name_y   
0                 NaN         NaN                NaN           NaN  \
1                 NaN         NaN                NaN           NaN   
2                 NaN         NaN                NaN           NaN   
3                 NaN         NaN                NaN           NaN   
4                 NaN         NaN                NaN           NaN   
...               ...         ...                ...           ...   
3029659           NaN         NaN                NaN           NaN   
3029660           NaN         NaN                NaN           NaN   
3029661           NaN         NaN                NaN           NaN   
3029662           NaN         NaN                NaN           NaN   
3029663           NaN         NaN                NaN           NaN   

         locale_regional  type_regional  transferred_regional  
0                    NaN            NaN                   NaN  
1                    NaN            NaN                   NaN  
2                    NaN            NaN                   NaN  
3                    NaN            NaN                   NaN  
4                    NaN            NaN                   NaN  
...                  ...            ...                   ...  
3029659              NaN            NaN                   NaN  
3029660              NaN            NaN                   NaN  
3029661              NaN            NaN                   NaN  
3029662              NaN            NaN                   NaN  
3029663              NaN            NaN                   NaN  

[3029664 rows x 18 columns]

In [125]:
loja_salles6 = loja_salles5.merge(feriado_National, left_on=['date'], right_on=['date'], how='left')

loja_salles6.drop(labels=['locale_name_x', 'locale_name_y', 'locale_name','locale_name'], axis=1, inplace=True)


In [126]:
holidays_events['locale']

0         Local
1      Regional
2         Local
3         Local
4         Local
         ...   
345    National
346    National
347    National
348    National
349    National
Name: locale, Length: 350, dtype: object

In [127]:
loja_salles6.loc[:, ['locale_local', 'type_local',
       'transferred_local', 'locale_regional', 'type_regional',
       'transferred_regional', 'locale_national', 'type_national',
       'transferred_national']] = loja_salles6.loc[:, ['locale_local', 'type_local',
       'transferred_local', 'locale_regional', 'type_regional',
       'transferred_regional', 'locale_national', 'type_national',
       'transferred_national']].fillna(-1)


In [128]:
#Não existe duplicada na verdade
loja_salles6[loja_salles6.duplicated()]

date  store_nbr                      family       sales   
23827   2014-12-26          1                  AUTOMOTIVE     1.00000  \
23829   2014-12-26          1                   BABY CARE     0.00000   
23831   2014-12-26          1                      BEAUTY     3.00000   
23833   2014-12-26          1                   BEVERAGES  1764.00000   
23835   2014-12-26          1                       BOOKS     0.00000   
...            ...        ...                         ...         ...   
3020976 2016-05-08          9                     POULTRY   590.37103   
3020978 2016-05-08          9              PREPARED FOODS   110.06000   
3020980 2016-05-08          9                     PRODUCE  1782.35200   
3020982 2016-05-08          9  SCHOOL AND OFFICE SUPPLIES     7.00000   
3020984 2016-05-08          9                     SEAFOOD    11.39400   

         onpromotion   city      state type  cluster  transactions   
23827              0  Quito  Pichincha    D       13        1295.0  \
23829              0  Quito  Pichincha    D       13        1295.0   
23831              0  Quito  Pichincha    D       13        1295.0   
23833              4  Quito  Pichincha    D       13        1295.0   
23835              0  Quito  Pichincha    D       13        1295.0   
...              ...    ...        ...  ...      ...           ...   
3020976            0  Quito  Pichincha    B        6        2389.0   
3020978            0  Quito  Pichincha    B        6        2389.0   
3020980            1  Quito  Pichincha    B        6        2389.0   
3020982            0  Quito  Pichincha    B        6        2389.0   
3020984            1  Quito  Pichincha    B        6        2389.0   

         locale_local  type_local  transferred_local  locale_regional   
23827            -1.0        -1.0               -1.0             -1.0  \
23829            -1.0        -1.0               -1.0             -1.0   
23831            -1.0        -1.0               -1.0             -1.0   
23833            -1.0        -1.0               -1.0             -1.0   
23835            -1.0        -1.0               -1.0             -1.0   
...               ...         ...                ...              ...   
3020976          -1.0        -1.0               -1.0             -1.0   
3020978          -1.0        -1.0               -1.0             -1.0   
3020980          -1.0        -1.0               -1.0             -1.0   
3020982          -1.0        -1.0               -1.0             -1.0   
3020984          -1.0        -1.0               -1.0             -1.0   

         type_regional  transferred_regional  locale_national  type_national   
23827             -1.0                  -1.0              1.0            0.0  \
23829             -1.0                  -1.0              1.0            0.0   
23831             -1.0                  -1.0              1.0            0.0   
23833             -1.0                  -1.0              1.0            0.0   
23835             -1.0                  -1.0              1.0            0.0   
...                ...                   ...              ...            ...   
3020976           -1.0                  -1.0              1.0            0.0   
3020978           -1.0                  -1.0              1.0            0.0   
3020980           -1.0                  -1.0              1.0            0.0   
3020982           -1.0                  -1.0              1.0            0.0   
3020984           -1.0                  -1.0              1.0            0.0   

         transferred_national  
23827                     0.0  
23829                     0.0  
23831                     0.0  
23833                     0.0  
23835                     0.0  
...                       ...  
3020976                   0.0  
3020978                   0.0  
3020980                   0.0  
3020982                   0.0  
3020984                   0.0  

[7392 rows x 19 columns]

In [129]:
loja_salles6['locale_regional'].unique()

array([-1.,  2.])

In [130]:
loja_salles6.isnull().sum()

date                         0
store_nbr                    0
family                       0
sales                    28512
onpromotion                  0
city                         0
state                        0
type                         0
cluster                      0
transactions            274593
locale_local                 0
type_local                   0
transferred_local            0
locale_regional              0
type_regional                0
transferred_regional         0
locale_national              0
type_national                0
transferred_national         0
dtype: int64

In [131]:
loja_salles6[loja_salles6['type_local'] == 1]

Empty DataFrame
Columns: [date, store_nbr, family, sales, onpromotion, city, state, type, cluster, transactions, locale_local, type_local, transferred_local, locale_regional, type_regional, transferred_regional, locale_national, type_national, transferred_national]
Index: []

### Esses 28512 registros nullos são na verdade os dados de treino de teste, não podemos apagar.

In [132]:
loja_salles6[loja_salles6['sales'].isnull()]

date  store_nbr                      family  sales  onpromotion   
55704   2017-08-16          1                  AUTOMOTIVE    NaN            0  \
55705   2017-08-16          1                   BABY CARE    NaN            0   
55706   2017-08-16          1                      BEAUTY    NaN            2   
55707   2017-08-16          1                   BEVERAGES    NaN           20   
55708   2017-08-16          1                       BOOKS    NaN            0   
...            ...        ...                         ...    ...          ...   
3036787 2017-08-31          9                     POULTRY    NaN            1   
3036788 2017-08-31          9              PREPARED FOODS    NaN            0   
3036789 2017-08-31          9                     PRODUCE    NaN            1   
3036790 2017-08-31          9  SCHOOL AND OFFICE SUPPLIES    NaN            9   
3036791 2017-08-31          9                     SEAFOOD    NaN            0   

          city      state type  cluster  transactions  locale_local   
55704    Quito  Pichincha    D       13           NaN          -1.0  \
55705    Quito  Pichincha    D       13           NaN          -1.0   
55706    Quito  Pichincha    D       13           NaN          -1.0   
55707    Quito  Pichincha    D       13           NaN          -1.0   
55708    Quito  Pichincha    D       13           NaN          -1.0   
...        ...        ...  ...      ...           ...           ...   
3036787  Quito  Pichincha    B        6           NaN          -1.0   
3036788  Quito  Pichincha    B        6           NaN          -1.0   
3036789  Quito  Pichincha    B        6           NaN          -1.0   
3036790  Quito  Pichincha    B        6           NaN          -1.0   
3036791  Quito  Pichincha    B        6           NaN          -1.0   

         type_local  transferred_local  locale_regional  type_regional   
55704          -1.0               -1.0             -1.0           -1.0  \
55705          -1.0               -1.0             -1.0           -1.0   
55706          -1.0               -1.0             -1.0           -1.0   
55707          -1.0               -1.0             -1.0           -1.0   
55708          -1.0               -1.0             -1.0           -1.0   
...             ...                ...              ...            ...   
3036787        -1.0               -1.0             -1.0           -1.0   
3036788        -1.0               -1.0             -1.0           -1.0   
3036789        -1.0               -1.0             -1.0           -1.0   
3036790        -1.0               -1.0             -1.0           -1.0   
3036791        -1.0               -1.0             -1.0           -1.0   

         transferred_regional  locale_national  type_national   
55704                    -1.0             -1.0           -1.0  \
55705                    -1.0             -1.0           -1.0   
55706                    -1.0             -1.0           -1.0   
55707                    -1.0             -1.0           -1.0   
55708                    -1.0             -1.0           -1.0   
...                       ...              ...            ...   
3036787                  -1.0             -1.0           -1.0   
3036788                  -1.0             -1.0           -1.0   
3036789                  -1.0             -1.0           -1.0   
3036790                  -1.0             -1.0           -1.0   
3036791                  -1.0             -1.0           -1.0   

         transferred_national  
55704                    -1.0  
55705                    -1.0  
55706                    -1.0  
55707                    -1.0  
55708                    -1.0  
...                       ...  
3036787                  -1.0  
3036788                  -1.0  
3036789                  -1.0  
3036790                  -1.0  
3036791                  -1.0  

[28512 rows x 19 columns]

### Trandormando a coluna date de OIL em DATA

In [133]:
oil['date'] = pd.to_datetime(oil['date'])

In [134]:
oil[oil['dcoilwtico'].isna()].head()

date  dcoilwtico
0   2013-01-01         NaN
14  2013-01-21         NaN
34  2013-02-18         NaN
63  2013-03-29         NaN
104 2013-05-27         NaN

In [135]:
oil['dcoilwtico'] = oil['dcoilwtico'].interpolate(method='linear')

#Orientação Chat
# A interpolação é uma técnica usada para preencher valores ausentes (NaN) com base nos valores conhecidos próximos.
# No caso da interpolação linear, os valores ausentes são preenchidos seguindo uma linha reta entre os valores conhecidos antes e depois do NaN.
# Isso ajuda a estimar os preços do petróleo para as datas em que não há dados disponíveis, com base na tendência dos preços próximos.


In [136]:
oil[oil['date'] == '2013-01-21']

date  dcoilwtico
14 2013-01-21       95.85

In [137]:
oil[oil['dcoilwtico'].isna()].head()

date  dcoilwtico
0 2013-01-01         NaN

In [138]:
# A interpolação precisa de valores antes e depois para fazer a suposição e substituição de valores... que foi  escolhido para preencher os valores vazios do preço do pretoleo. Como a data 2013-01-01 não tem valores antecedentes , continuou vazio. logo uso o ffil, que preenche valores de acordo com preenchimento para trás. 
oil['dcoilwtico'] = oil['dcoilwtico'].fillna(method='bfill')  # Preenche para trás


In [139]:
oil[oil['date'] == '2013-01-01']

date  dcoilwtico
0 2013-01-01       93.14

In [140]:
# Sem valores nullos
oil[oil['dcoilwtico'].isna()].head()

Empty DataFrame
Columns: [date, dcoilwtico]
Index: []

#### Preço diário do petróleo. Inclui valores durante os períodos de dados de treinamento e teste. (O Equador é um país dependente do petróleo e sua saúde econômica é altamente vulnerável a choques nos preços do petróleo.)


Estou colocando no nosso data_frame o valor do petroleo 

In [141]:
loja_salles7 = loja_salles6.merge(oil,on='date',how='left')





### Não se esqueça que todos os -1 são colunasa vazias, logo se estão todas as colunas com -1, não houve feriado. 

In [142]:
loja_salles7[loja_salles7['sales'] != 0]

date  store_nbr                      family     sales   
33      2013-01-02          1                  AUTOMOTIVE     2.000  \
35      2013-01-02          1                      BEAUTY     2.000   
36      2013-01-02          1                   BEVERAGES  1091.000   
38      2013-01-02          1                BREAD/BAKERY   470.652   
40      2013-01-02          1                    CLEANING  1060.000   
...            ...        ...                         ...       ...   
3036787 2017-08-31          9                     POULTRY       NaN   
3036788 2017-08-31          9              PREPARED FOODS       NaN   
3036789 2017-08-31          9                     PRODUCE       NaN   
3036790 2017-08-31          9  SCHOOL AND OFFICE SUPPLIES       NaN   
3036791 2017-08-31          9                     SEAFOOD       NaN   

         onpromotion   city      state type  cluster  transactions   
33                 0  Quito  Pichincha    D       13        2111.0  \
35                 0  Quito  Pichincha    D       13        2111.0   
36                 0  Quito  Pichincha    D       13        2111.0   
38                 0  Quito  Pichincha    D       13        2111.0   
40                 0  Quito  Pichincha    D       13        2111.0   
...              ...    ...        ...  ...      ...           ...   
3036787            1  Quito  Pichincha    B        6           NaN   
3036788            0  Quito  Pichincha    B        6           NaN   
3036789            1  Quito  Pichincha    B        6           NaN   
3036790            9  Quito  Pichincha    B        6           NaN   
3036791            0  Quito  Pichincha    B        6           NaN   

         locale_local  type_local  transferred_local  locale_regional   
33               -1.0        -1.0               -1.0             -1.0  \
35               -1.0        -1.0               -1.0             -1.0   
36               -1.0        -1.0               -1.0             -1.0   
38               -1.0        -1.0               -1.0             -1.0   
40               -1.0        -1.0               -1.0             -1.0   
...               ...         ...                ...              ...   
3036787          -1.0        -1.0               -1.0             -1.0   
3036788          -1.0        -1.0               -1.0             -1.0   
3036789          -1.0        -1.0               -1.0             -1.0   
3036790          -1.0        -1.0               -1.0             -1.0   
3036791          -1.0        -1.0               -1.0             -1.0   

         type_regional  transferred_regional  locale_national  type_national   
33                -1.0                  -1.0             -1.0           -1.0  \
35                -1.0                  -1.0             -1.0           -1.0   
36                -1.0                  -1.0             -1.0           -1.0   
38                -1.0                  -1.0             -1.0           -1.0   
40                -1.0                  -1.0             -1.0           -1.0   
...                ...                   ...              ...            ...   
3036787           -1.0                  -1.0             -1.0           -1.0   
3036788           -1.0                  -1.0             -1.0           -1.0   
3036789           -1.0                  -1.0             -1.0           -1.0   
3036790           -1.0                  -1.0             -1.0           -1.0   
3036791           -1.0                  -1.0             -1.0           -1.0   

         transferred_national  dcoilwtico  
33                       -1.0       93.14  
35                       -1.0       93.14  
36                       -1.0       93.14  
38                       -1.0       93.14  
40                       -1.0       93.14  
...                       ...         ...  
3036787                  -1.0       47.26  
3036788                  -1.0       47.26  
3036789                  -1.0       47.26  
3036790                  -1.0       47.26 

In [143]:
loja_salles7['transferred_national'].unique()

array([ 0., -1.,  1.])

In [144]:
# Substituir -1 e NaN por 0
lista_feriado = loja_salles7[['type_local','transferred_local','type_regional','transferred_regional','type_national','transferred_national']].fillna(0).replace(-1, 0)

# Somando todas as colunas
tipo_feriado = lista_feriado[['type_local', 'type_regional', 'type_national']].sum(axis=1)
transferido = lista_feriado[['transferred_local', 'transferred_regional', 'transferred_national']].sum(axis=1)

# Se a soma for 0, então é feriado (0) | Se não for 0, é dia normal (1)
loja_salles7['Feriado'] = ((tipo_feriado + transferido) != 0).astype(int)



In [145]:
loja_salles7.columns

Index(['date', 'store_nbr', 'family', 'sales', 'onpromotion', 'city', 'state',
       'type', 'cluster', 'transactions', 'locale_local', 'type_local',
       'transferred_local', 'locale_regional', 'type_regional',
       'transferred_regional', 'locale_national', 'type_national',
       'transferred_national', 'dcoilwtico', 'Feriado'],
      dtype='object')

In [146]:
loja_salles7['locale_local'].unique()

array([-1.,  0.])

In [147]:
loja_salles7.head()

date  store_nbr      family  sales  onpromotion   city      state   
0 2013-01-01          1  AUTOMOTIVE    0.0            0  Quito  Pichincha  \
1 2013-01-01          1   BABY CARE    0.0            0  Quito  Pichincha   
2 2013-01-01          1      BEAUTY    0.0            0  Quito  Pichincha   
3 2013-01-01          1   BEVERAGES    0.0            0  Quito  Pichincha   
4 2013-01-01          1       BOOKS    0.0            0  Quito  Pichincha   

  type  cluster  transactions  ...  type_local  transferred_local   
0    D       13           NaN  ...        -1.0               -1.0  \
1    D       13           NaN  ...        -1.0               -1.0   
2    D       13           NaN  ...        -1.0               -1.0   
3    D       13           NaN  ...        -1.0               -1.0   
4    D       13           NaN  ...        -1.0               -1.0   

   locale_regional  type_regional  transferred_regional  locale_national   
0             -1.0           -1.0                  -1.0              1.0  \
1             -1.0           -1.0                  -1.0              1.0   
2             -1.0           -1.0                  -1.0              1.0   
3             -1.0           -1.0                  -1.0              1.0   
4             -1.0           -1.0                  -1.0              1.0   

   type_national  transferred_national  dcoilwtico  Feriado  
0            0.0                   0.0       93.14        0  
1            0.0                   0.0       93.14        0  
2            0.0                   0.0       93.14        0  
3            0.0                   0.0       93.14        0  
4            0.0                   0.0       93.14        0  

[5 rows x 21 columns]

In [148]:
# Retirando colunas desnecessarias , já que o resumo delas foi para a coluna feriado.
loja_salles8 = loja_salles7.drop(columns=['locale_local', 'type_local',
       'transferred_local', 'locale_regional', 'type_regional',
       'transferred_regional', 'locale_national', 'type_national',
       'transferred_national'])


In [149]:
# colocando novamente a coluna locale, para saber se o feriado era nacional ,regional ou local.
loja_salles8['locale'] = feriado['locale']
loja_salles8['locale'].fillna(-1,inplace=True)

### Pretendo transformar as dados categoricos de state,city e family em números. para facilitar no aprendizado de maquina.

In [150]:
# Isso transforma os dados em númericos para facilitar( ou boleanos no caso)
loja_salles9 = pd.get_dummies(loja_salles8[['state','family','city','type']],drop_first=True,prefix=['state','family','city','type'])

In [151]:
# Criando novo DataFrame com valores transformados em binarios e retirando colunas repetidas como state,family ,city
loja_salles10 = pd.concat([loja_salles8.drop(columns=['state','family','city','type']),loja_salles9],axis=1)

In [152]:
# Vou deixar essa coluna afim de que as lojas possam ser diferenciadas uma das outras
loja_salles10['store_nbr']

0          1
1          1
2          1
3          1
4          1
          ..
3036787    9
3036788    9
3036789    9
3036790    9
3036791    9
Name: store_nbr, Length: 3036792, dtype: int64

In [153]:
loja_salles10['locale'].isnull().sum()

0

# Os salários no setor público são pagos quinzenalmente no dia 15 e no último dia do mês. As vendas nos supermercados podem ser afetadas por isso.

In [154]:
# # Orientação ChatGPT
# Função para verificar se o dia é o 15 ou o último dia do mês
def is_payment_day(date):
    if date.day == 15 or date == date.replace(day=1) + pd.DateOffset(months=1, days=-1):
        return 1
    else:
        return 0

# Aplicando a função para criar a nova coluna
loja_salles10['payment_day'] = loja_salles10['date'].apply(is_payment_day)

In [155]:
loja_salles10['locale'].isnull().sum()

0

# Um terremoto de magnitude 7,8 atingiu o Equador em 16 de abril de 2016. As pessoas se reuniram nos esforços de socorro doando água e outros produtos de primeira necessidade, o que afetou muito as vendas dos supermercados por várias semanas após o terremoto.


# Orientação ChatGPT
Alternativa: Colocar uma escala de impacto (caso o efeito seja gradual):
Se você achar que o impacto nas vendas não é abrupto e sim gradual, pode considerar usar valores diferentes para o impacto. Ao invés de apenas 0 ou 1, você poderia criar uma escala de impacto, como:

0: Antes do terremoto.
1: Durante o terremoto e imediato após o evento.
0.5: Em um período intermediário de recuperação.
0: Após o impacto significativo ter passado.
Isso poderia representar uma tendência de diminuição do impacto ao longo do tempo.

In [156]:
# Definindo o início e fim do impacto (16 de abril de 2016 até 16 de julho de 2016)
start_date = pd.to_datetime('2016-04-16')
end_date = start_date + pd.DateOffset(months=3)  # Três meses após o terremoto

# Função para verificar se a data está no período de impacto
def is_earthquake_impact(date):
    if start_date <= date <= end_date:
        return 1
    else:
        return 0
    


# Criando a nova coluna 'earthquake_impact'
loja_salles10['earthquake_impact'] = loja_salles10['date'].apply(is_earthquake_impact)

In [157]:
loja_salles10['locale'].isnull().sum()

0

In [158]:
loja_salles10[loja_salles10['dcoilwtico'].isna()]

date  store_nbr   sales  onpromotion  cluster  transactions   
132     2013-01-05          1     5.0            0       13        1509.0  \
133     2013-01-05          1     0.0            0       13        1509.0   
134     2013-01-05          1     3.0            0       13        1509.0   
135     2013-01-05          1  1160.0            0       13        1509.0   
136     2013-01-05          1     0.0            0       13        1509.0   
...            ...        ...     ...          ...      ...           ...   
3036655 2017-08-27          9     NaN            0        6           NaN   
3036656 2017-08-27          9     NaN            0        6           NaN   
3036657 2017-08-27          9     NaN            2        6           NaN   
3036658 2017-08-27          9     NaN            9        6           NaN   
3036659 2017-08-27          9     NaN            0        6           NaN   

         dcoilwtico  Feriado  locale  state_Bolivar  ...  city_Quito   
132             NaN        1     0.0          False  ...        True  \
133             NaN        1     0.0          False  ...        True   
134             NaN        1     0.0          False  ...        True   
135             NaN        1     1.0          False  ...        True   
136             NaN        1     1.0          False  ...        True   
...             ...      ...     ...            ...  ...         ...   
3036655         NaN        0    -1.0          False  ...        True   
3036656         NaN        0    -1.0          False  ...        True   
3036657         NaN        0    -1.0          False  ...        True   
3036658         NaN        0    -1.0          False  ...        True   
3036659         NaN        0    -1.0          False  ...        True   

         city_Riobamba  city_Salinas  city_Santo Domingo  type_B  type_C   
132              False         False               False   False   False  \
133              False         False               False   False   False   
134              False         False               False   False   False   
135              False         False               False   False   False   
136              False         False               False   False   False   
...                ...           ...                 ...     ...     ...   
3036655          False         False               False    True   False   
3036656          False         False               False    True   False   
3036657          False         False               False    True   False   
3036658          False         False               False    True   False   
3036659          False         False               False    True   False   

         type_D  type_E  payment_day  earthquake_impact  
132        True   False            0                  0  
133        True   False            0                  0  
134        True   False            0                  0  
135        True   False            0                  0  
136        True   False            0                  0  
...         ...     ...          ...                ...  
3036655   False   False            0                  0  
3036656   False   False            0                  0  
3036657   False   False            0                  0  
3036658   False   False            0                  0  
3036659   False   False            0                  0  

[869880 rows x 83 columns]

Com uma quantidade tão grande de valores vazios na coluna "dcoilwtico", não posso simplesmente usar a interpolação, isso criaria vieses... por tanto, irei usar apenas a substiuição por bffill, tendo em mente que os o preço do pretoleo não muda todo dia, mas sim de forma gradual, logo isso não vai interferir muito. 

Por quê Essa opção ?

Isso é mais fiel à realidade.

O mercado de petróleo geralmente não tem alterações diárias.

Evita colocar dados falsos que o modelo vai achar que são reais.


In [159]:
loja_salles10['dcoilwtico'] = loja_salles10['dcoilwtico'].fillna(method='ffill')

In [160]:
# Verificando o restante das colunas vazias... o Salles é o treinamento então é normal
# Já a transactions, pensar o que fazer sobre, usar interpolação? ou media dos valores, mediana?.... vamos verificar. 
loja_salles10.isna().sum()[loja_salles10.isna().sum() > 0]

sales            28512
transactions    274593
dtype: int64

In [161]:
# Verificando quais as unicas colunas que possuem valores vazios.
loja_salles10.columns[loja_salles10.isna().any()]

Index(['sales', 'transactions'], dtype='object')

In [162]:
loja_salles10.isna().sum()[['sales','transactions']]

sales            28512
transactions    274593
dtype: int64

In [163]:
# Não foi encontrado nenhuma outro melhor forma de descobrir a importancia da coluna transactions, e por suas muitas linha vazias, deci remove-la.
loja_salles10 = loja_salles10.drop(columns='transactions',axis=1).copy()

In [164]:
loja_salles10.columns[loja_salles10.isna().any()]

Index(['sales'], dtype='object')

In [165]:
# é necessario uma transformação da coluna data para que se torne númerica, tendo em mente que os modelos não lidam como type DATATIME, logo iremos usar o metodo para transformar a data em númerico, separando dias , meses e anos. . exemplo:

loja_salles10['day'] = loja_salles10['date'].dt.day
loja_salles10['month'] = loja_salles10['date'].dt.month
loja_salles10['year'] = loja_salles10['date'].dt.year

loja_salles10 = loja_salles10.drop(columns='date')

'''
2. Extrair componentes da data (ano, mês, dia, dia da semana, etc.):
Quando usar:

Quando você sabe que a variação sazonal (como mês ou dia da semana) tem impacto no seu modelo.
Quando você deseja que o modelo tenha acesso a informações explícitas sobre o tempo, como prever tendências mensais ou sazonais.

'''




'\n2. Extrair componentes da data (ano, mês, dia, dia da semana, etc.):\nQuando usar:\n\nQuando você sabe que a variação sazonal (como mês ou dia da semana) tem impacto no seu modelo.\nQuando você deseja que o modelo tenha acesso a informações explícitas sobre o tempo, como prever tendências mensais ou sazonais.\n\n'

In [166]:
loja_salles10.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3036792 entries, 0 to 3036791
Data columns (total 84 columns):
 #   Column                                Dtype  
---  ------                                -----  
 0   store_nbr                             int64  
 1   sales                                 float64
 2   onpromotion                           int64  
 3   cluster                               int64  
 4   dcoilwtico                            float64
 5   Feriado                               int32  
 6   locale                                float64
 7   state_Bolivar                         bool   
 8   state_Chimborazo                      bool   
 9   state_Cotopaxi                        bool   
 10  state_El Oro                          bool   
 11  state_Esmeraldas                      bool   
 12  state_Guayas                          bool   
 13  state_Imbabura                        bool   
 14  state_Loja                            bool   
 15  state_Los Rios 

================================================

In [167]:
# importando bibliotecas de aprendizado de maquina somente para aprendizado de regressão, já que estamos querendo prever um valor continuo

# Bibliotecas que se adptam melhor a modelos de regressão
from sklearn.linear_model import LinearRegression, LogisticRegression

# Bibliotecas que utilizam modelos de arvores , tanto para regressão quanto para classificação
from sklearn.tree import  DecisionTreeRegressor

#
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor,HistGradientBoostingRegressor

# Bibliotecas de vizinhos, utilizam a aproximação de uma da outra para classificar, ou fazer uma indução de regreção
from sklearn.neighbors import  KNeighborsRegressor

# Bibliotecas para melhorar a acuracia dos dados, e compreender melhor sua qualidade( no caso a qualidade das previsões do aprendizado de maquina)
from sklearn.metrics import accuracy_score, precision_score, recall_score, mean_squared_error

# Biblioteca utilizada para treinar a maquina com os dados disponiveis. 
from sklearn.model_selection import train_test_split, cross_val_score

# O GridSearchCV é usado para encontrarmos os melhores parametros para nosso modelo de regressão. 
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

from sklearn.svm import SVR

# modelo de árvore para regressão
from sklearn.tree import DecisionTreeRegressor

#Utilizado para limpar variáveis antigas ocupando RAM
import gc





### Separando os dados de treino dos dados de teste.

In [168]:
# pegando os valores de teste retirando os valores de treino
teste = loja_salles10[loja_salles10['sales'].isna()]

In [169]:
teste.head()

store_nbr  sales  onpromotion  cluster  dcoilwtico  Feriado  locale   
55704          1    NaN            0       13        46.8        0    -1.0  \
55705          1    NaN            0       13        46.8        0    -1.0   
55706          1    NaN            2       13        46.8        0    -1.0   
55707          1    NaN           20       13        46.8        0    -1.0   
55708          1    NaN            0       13        46.8        0    -1.0   

       state_Bolivar  state_Chimborazo  state_Cotopaxi  ...   
55704          False             False           False  ...  \
55705          False             False           False  ...   
55706          False             False           False  ...   
55707          False             False           False  ...   
55708          False             False           False  ...   

       city_Santo Domingo  type_B  type_C  type_D  type_E  payment_day   
55704               False   False   False    True   False            0  \
55705               False   False   False    True   False            0   
55706               False   False   False    True   False            0   
55707               False   False   False    True   False            0   
55708               False   False   False    True   False            0   

       earthquake_impact  day  month  year  
55704                  0   16      8  2017  
55705                  0   16      8  2017  
55706                  0   16      8  2017  
55707                  0   16      8  2017  
55708                  0   16      8  2017  

[5 rows x 84 columns]

In [170]:
# pegando os valores de treino retirando os valores de teste
treino = loja_salles10[loja_salles10['sales'].notna()]

In [171]:
# Verificando se há valores nullos
treino.columns.isna().sum().any()

False

In [172]:
treino.head()

store_nbr  sales  onpromotion  cluster  dcoilwtico  Feriado  locale   
0          1    0.0            0       13       93.14        0     0.0  \
1          1    0.0            0       13       93.14        0     2.0   
2          1    0.0            0       13       93.14        0     0.0   
3          1    0.0            0       13       93.14        0     0.0   
4          1    0.0            0       13       93.14        0     0.0   

   state_Bolivar  state_Chimborazo  state_Cotopaxi  ...  city_Santo Domingo   
0          False             False           False  ...               False  \
1          False             False           False  ...               False   
2          False             False           False  ...               False   
3          False             False           False  ...               False   
4          False             False           False  ...               False   

   type_B  type_C  type_D  type_E  payment_day  earthquake_impact  day  month   
0   False   False    True   False            0                  0    1      1  \
1   False   False    True   False            0                  0    1      1   
2   False   False    True   False            0                  0    1      1   
3   False   False    True   False            0                  0    1      1   
4   False   False    True   False            0                  0    1      1   

   year  
0  2013  
1  2013  
2  2013  
3  2013  
4  2013  

[5 rows x 84 columns]

existe algumas colunas que continuam com valores de classificação em vez de binarios. não usarei o get_dummies, para verificar a acuracia.

* store_nbr

* cluster

* locale



In [173]:
# Separando os dados de teste, para que no X esteja os dados para aprendizado, e no y as respostas(no caso o SALLES)
X = treino.drop(columns='sales',axis=1)
y = treino['sales']

In [174]:
# Explicação:
# Estamos dividindo 30% para testes, e 70% para treinamento.
# os que começam com x e estão em primeiro são os 30% e o segundo que começa com x tem os 70% dos dados de treinamento
# já os que começam com y, são as respostas dos respectivos testes. 


X_train,X_teste,y_train,y_teste = train_test_split(X,y,test_size=0.3,random_state=42)

In [ ]:
parametros_tree_regressor = {
        'max_depth': [5, 10],  
        'min_samples_split': [5, 10],  
        'min_samples_leaf': [5, 10] 
    }

parametros_vecto_regressor ={
    'C': [0.1, 1, 10],
    'epsilon': [0.01, 0.1, 1],
    'kernel': ['linear', 'rbf']
}

parametros_gradient_boostin_regressor = {
    'n_estimators': [50, 100],  # Menos valores
    'learning_rate': [0.05, 0.1],  # Remova valores extremos
    'max_depth': [3, 5]  # Evite valores altos
}

parametros_hist_gradient_boosting = {
    'learning_rate': [0.01, 0.1],  # Menos valores
    'max_iter': [100, 200],
    'max_depth': [3, 5],
    'min_samples_leaf': [20, 50]
}


parametros_linear = {
    'fit_intercept': [True, False],
    'positive': [True, False]
}


parametros_knn_regressor = {
    'n_neighbors': [3, 5, 7],  
    'weights': ['uniform', 'distance'],  
    'algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute'],  
    'leaf_size': [20, 30, 40],  
    'p': [1, 2]  
}

In [179]:
def aplicando_hiper_parametros(modelo, parametros, X_train, y_train, X_teste, y_teste):

  

        # Criando o GridSearchCV
        grid_search = RandomizedSearchCV(modelo, parametros, cv=3, scoring='r2', n_jobs=1)

        # Treinando o GridSearchCV
        grid_search.fit(X_train, y_train)

        # Pegando o melhor modelo encontrado e seu melhor parametro, para possivel analise e comparação
        melhor_modelo = grid_search.best_estimator_
        melhores_parametros = grid_search.best_params_

        # Fazendo previsões 
        y_pred = melhor_modelo.predict(X_teste)

        # Calculando R² de treino e teste
        acuracia_train = round(melhor_modelo.score(X_train, y_train) * 100, 2)
        acuracia_test = round(melhor_modelo.score(X_teste, y_teste) * 100, 2)

        print("Melhores parâmetros:", melhores_parametros)
        print("Acuracia de treino:", acuracia_train)
        print("Acuracia de teste:", acuracia_test)



    #return melhor_modelo, y_pred, melhores_parametros


In [180]:
# Aprendizado de maquina DecisionTreeRegressor()
aplicando_hiper_parametros(DecisionTreeRegressor(),parametros_tree_regressor,X_train, y_train, X_teste, y_teste)

c:\Users\Thiago\Documents\devenvs\ambiente_virtual2\lib\site-packages\sklearn\model_selection\_search.py:317: UserWarning: The total space of parameters 8 is smaller than n_iter=10. Running 8 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Melhores parâmetros: {'min_samples_split': 5, 'min_samples_leaf': 10, 'max_depth': 10}
Acuracia de treino: 85.17
Acuracia de teste: 85.63


In [ ]:
aplicando_hiper_parametros(1,HistGradientBoostingRegressor(),parametros_hist_gradient_boosting,X_train, y_train, X_teste, y_teste)

Melhores parâmetros: {'min_samples_leaf': 50, 'max_iter': 200, 'max_depth': 5, 'learning_rate': 0.1}
Acuracia de treino: 85.5
Acuracia de teste: 86.75


In [ ]:
# Aprendizado de maquina LinearRegression()
aplicando_hiper_parametros(LinearRegression(),parametros_linear,X_train, y_train, X_teste, y_teste)

c:\Users\Thiago\Documents\devenvs\ambiente_virtual2\lib\site-packages\sklearn\model_selection\_search.py:317: UserWarning: The total space of parameters 4 is smaller than n_iter=10. Running 4 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


In [ ]:
# Aprendizado de maquina KNeighborsRegressor()
#aplicando_hiper_parametros(KNeighborsRegressor(),parametros_knn_regressor,X_train, y_train, X_teste, y_teste)
